### PR TITLE
Fix race condition in extractAllGroups

### PR DESCRIPTION
**Pull request by alexey-milovidov**
Changelog category (leave one):
- Bug Fix

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix race condition in extractAllGroups* functions.

See https://clickhouse-test-reports.s3.yandex.net/11920/29a9d4187f569cc781cffbc712c1a3ddb3fcfb82/stress_test_(thread)/stderr.log

PS. We use `re2_st` instead of `re2` for reason.

### CODE REVIEW

File: src/Functions/MatchImpl.h [hunk 1]
```diff
@@ -126,7 +126,7 @@ struct MatchImpl
         {
             size_t size = offsets.size();
 
-            const auto & regexp = Regexps::get<like, true>(pattern);
+            auto regexp = Regexps::get<like, true>(pattern);
```
Comment (alexey-milovidov) [inline src/Functions/MatchImpl.h:129]: This is just for readability.

File: src/Functions/MatchImpl.h [hunk 2]
```diff
@@ -281,7 +281,7 @@ struct MatchImpl
         {
             size_t size = data.size() / n;
 
-            const auto & regexp = Regexps::get<like, true>(pattern);
+            auto regexp = Regexps::get<like, true>(pattern);
```
Comment (alexey-milovidov) [inline src/Functions/MatchImpl.h:284]: This is just for readability.

File: src/Functions/extractAllGroups.h [hunk 3]
```diff
@@ -82,7 +82,8 @@ class FunctionExtractAllGroups : public IFunction
             throw Exception("Length of 'needle' argument must be greater than 0.", ErrorCodes::BAD_ARGUMENTS);
 
         using StringPiece = typename Regexps::Regexp::StringPieceType;
-        const auto & regexp = Regexps::get<false, false>(needle)->getRE2();
+        auto holder = Regexps::get<false, false>(needle);
```
Comment (alexey-milovidov) [inline src/Functions/extractAllGroups.h:85]: Here was a bug. The ownership of an object from pool was not hold.

File: src/Functions/extractGroups.cpp [hunk 4]
```diff
@@ -61,7 +61,7 @@ class FunctionExtractGroups : public IFunction
         if (needle.empty())
             throw Exception(getName() + " length of 'needle' argument must be greater than 0.", ErrorCodes::BAD_ARGUMENTS);
 
-        const auto regexp = Regexps::get<false, false>(needle);
+        auto regexp = Regexps::get<false, false>(needle);
```
Comment (alexey-milovidov) [inline src/Functions/extractGroups.cpp:64]: This code was correct. The change is just for readability.

**Review by alesapin — APPROVED** [review]
LGTM

**Comment by alexey-milovidov** [comment]
The code was correct before this change: https://github.com/ClickHouse/ClickHouse/pull/11554/files#diff-89149b64c537652e192428ec0734b7d5L66

File: src/Functions/extractAllGroups.h (hunk 3 above)
Comment (Enmk) [reply src/Functions/extractAllGroups.h:85]: Thanks for fixing, could you please also add a comment to `Regexps::get` that caller has to store the result by value?

Also, I don't quite understand why const reference doesn't suffice here, as it supposed to extend lifetime of temporary object (the unnamed return value) as described in [The most important const](https://herbsutter.com/2008/01/01/gotw-88-a-candidate-for-the-most-important-const/l).

**Review by Enmk — APPROVED** [review]
all good, except for missing minor comment.

File: src/Functions/extractAllGroups.h (hunk 3 above)
Comment (alexey-milovidov) [reply src/Functions/extractAllGroups.h:85]: We have to hold
```
Regexps::get<false, false>(needle)
```

but we are using
```
Regexps::get<false, false>(needle)->getRE2()
```
instead.

Comment (alexey-milovidov) [reply src/Functions/extractAllGroups.h:85]: It returns `Pool::Pointer` so the reader has to ask: "what is pool?", "why do we have pool here?", "how do I take and return back objects into pool?".

I will add a comment.

Comment (alexey-milovidov) [reply src/Functions/extractAllGroups.h:85]: Added a comment in master.

Comment (Enmk) [reply src/Functions/extractAllGroups.h:85]: oh, I get it now, thanks

**Pull request merged**